flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java

```diff
@@ -0,0 +1,165 @@
+ * {@code TableSourceScan}. Meanwhile the output datatype of {@code TableSourceScan} has changed,
+ * which is the pattern of {@code grouping sets} + {@code the output of aggregate functions}.
+ *
+ * <p>Due to the complexity of aggregate, only limited aggregate functions are supported at present.
+ *
+ * <ul>
+ *   <li>Only support sum/min/max/count/avg(will convert to sum0 + count) aggregate function push
+ *       down.
+ *   <li>Only support simple group type, cube and roll up will not be pushed down.
+ *   <li>If expression is involved in aggregate or group by, e.g. max (col1 + col2) or group by
+ *       (col1 + col2), aggregate will not be pushed down.
+ *   <li>Window aggregate function will not be pushed down.
+ *   <li>Aggregate function with filter will not be pushed down.
+ * </ul>
+ *
+ * <p>For the above example inventory table, and we have the below test sql:
+ *
+ * <pre>{@code
+ * SELECT
+ *   sum(amount),
+ *   max(price),
+ *   avg(price),
+ *   count(distinct price),
+ *   name,
+ *   type
+ * FROM inventory
+ *   group by name, type
+ * }</pre>
+ *
+ * <p>Since there is a count(distinct) aggregate function in it, the entire computational semantics
+ * will change obviously. And the optimized plan as shown below. The local aggregate will not be
+ * pushed down in this scenario.
+ *
+ * <pre>{@code
+ * Calc(select=[EXPR$0, EXPR$1, EXPR$2, EXPR$3, name, type])
+ * +- HashAggregate(groupBy=[name, type], select=[name, type, Final_MIN(min$0) AS EXPR$0, Final_MIN(min$1) AS EXPR$1, Final_MIN(min$2) AS EXPR$2, Final_COUNT(count$3) AS EXPR$3])
+ *    +- Exchange(distribution=[hash[name, type]])
+ *       +- LocalHashAggregate(groupBy=[name, type], select=[name, type, Partial_MIN(EXPR$0) FILTER $g_1 AS min$0, Partial_MIN(EXPR$1) FILTER $g_1 AS min$1, Partial_MIN(EXPR$2) FILTER $g_1 AS min$2, Partial_COUNT(price) FILTER $g_0 AS count$3])
+ *          +- Calc(select=[name, type, price, EXPR$0, EXPR$1, EXPR$2, =(CASE(=($e, 0:BIGINT), 0:BIGINT, 1:BIGINT), 0) AS $g_0, =(CASE(=($e, 0:BIGINT), 0:BIGINT, 1:BIGINT), 1) AS $g_1])
+ *             +- HashAggregate(groupBy=[name, type, price, $e], select=[name, type, price, $e, Final_SUM(sum$0) AS EXPR$0, Final_MAX(max$1) AS EXPR$1, Final_AVG(sum$2, count$3) AS EXPR$2])
+ *                +- Exchange(distribution=[hash[name, type, price, $e]])
+ *                   +- LocalHashAggregate(groupBy=[name, type, price, $e], select=[name, type, price, $e, Partial_SUM(amount) AS sum$0, Partial_MAX(price_0) AS max$1, Partial_AVG(price_0) AS (sum$2, count$3)])
+ *                      +- Expand(projects=[name, type, amount, price, $e, price_0], projects=[{name, type, amount, price, 0 AS $e, price AS price_0}, {name, type, amount, null AS price, 1 AS $e, price AS price_0}])
+ *                         +- TableSourceScan(table=[[inventory, project=[name, type, amount, price]]], fields=[name, type, amount, price])
+ * }</pre>
```

Comment:
I also think this is not a good example, and we don't need to mention it. Some databases can support count(distinct)

Reply:
Have fixed, change to use aggregate with filter as an unable push down demo.